init.lua

```diff
@@ -1,55 +1,60 @@
-local original_pos = {}
+local original_state = {}
 
 minetest.register_privilege("watch", {
 	description = "Player can watch other players",
 	give_to_singleplayer = false,
 	give_to_admin = true,
 })
 
-local function toggle_hud_flags(player, bool)
+local function turn_off_hud_flags(player)
 	local flags = player:hud_get_flags()
 	local new_hud_flags = {}
 
 	for flag in pairs(flags) do
-		new_hud_flags[flag] = bool
+		new_hud_flags[flag] = false
 	end
 
 	player:hud_set_flags(new_hud_flags)
 end
 
-local function unwatching(name)
+local function detach(name)
+	-- nothing to do
+	if not player_api.player_attached[name] then return end
+
 	local watcher = minetest.get_player_by_name(name)
-	local privs = minetest.get_player_privs(name)
+	if not watcher then return end -- shouldn't ever happen
 
-	if watcher and default.player_attached[name] == true then
-		watcher:set_detach()
-		player_api.player_attached[name] = false
-		watcher:set_eye_offset(vector.new(), vector.new())
-		watcher:set_nametag_attributes({color = {a = 255, r = 255, g = 255, b = 255}})
+	watcher:set_detach()
+	player_api.player_attached[name] = false
+	watcher:set_eye_offset(vector.new(), vector.new())
 
-		toggle_hud_flags(watcher, true)
+	local saved_state = original_state[name]
+	-- nothing else to do
+	if not saved_state then return end
 
-		watcher:set_properties({
-			visual_size = {x = 1, y = 1},
-			makes_footstep_sound = true,
-			collisionbox = {-0.3, 0.0, -0.3, 0.3, 1.7, 0.3}
-		})
+	watcher:set_nametag_attributes({ color = saved_state.nametag.color, bgcolor = saved_state.nametag.bgcolor })
+	watcher:hud_set_flags(saved_state.hud_flags)
+	watcher:set_properties({
+		visual_size = saved_state.visual_size,
+		makes_footstep_sound = true,
+		collisionbox = saved_state.collisionbox,
+	})
 
-		if not privs.interact and privs.watch == true then
-			privs.interact = true
-			minetest.set_player_privs(name, privs)
-		end
+	local privs = minetest.get_player_privs(name)
+	if not privs.interact and privs.watch then
```

Comment:
Why does it depend on `watch`? If they're attached to a player, detach it properly under all circumstances.

Reply:
yes, exactly what I asked myself too. Have different procedure in 'invite' branch. The thing to check is if the player had `interact` prior to being attached, if he has `watch` shouldn't matter.